content/match-types-spec.md

```diff
@@ -0,0 +1,579 @@
+* `C ⋔ D` if `C` is `sealed` without anonymous child and `Ci ⋔ D` for all direct children `Ci` of `C`
+* `C ⋔ D` if `D` is `sealed` without anonymous child and `C ⋔ Di` for all direct children `Di` of `D`
+
+We can now define `⋔` for *types*.
+
+For arbitrary types `X` and `Y`, we define `X ⋔ Y` as `⌈X⌉ ⋔ ⌈Y⌉`.
+
+Two simple types `S` and `T` are provably disjoint if there is a finite derivation tree for `S ⋔ T` using the following rules.
+Most rules go by pair, which makes the whole relation symmetric:
+
+* `Nothing` is disjoint from everything (including itself):
+  * `Nothing ⋔ T`
+  * `S ⋔ Nothing`
+* A union type is disjoint from another type if both of its parts are disjoint from that type:
+  * `S ⋔ T1 | T2` if `S ⋔ T1` and `S ⋔ T2`
+  * `S1 | S2 ⋔ T` if `S1 ⋔ T` and `S2 ⋔ T`
+* An intersection type is disjoint from another type if at least one of its parts is disjoint from that type:
+  * `S ⋔ T1 & T2` if `S ⋔ T1` or `S ⋔ T2`
+  * `S1 & S2 ⋔ T` if `S1 ⋔ T` or `S1 ⋔ T`
+* A type lambda is disjoint from any other type that is not a type lambda with the same number of parameters:
+  * `[...ai] =>> S1 ⋔ q.D.y`
+  * `[...ai] =>> S1 ⋔ d`
+  * `[...ai] =>> S1 ⋔ q.D[...Ti]`
+  * `p.C.x ⋔ [...bi] =>> T1`
+  * `c ⋔ [...bi] =>> T1`
+  * `p.C[...Si] ⋔ [...bi] =>> T1`
+  * `[a1, ..., an] =>> S1 ⋔ [b1, ..., bm] =>> T1` if `m != n`
+* Two type lambdas with the same number of type parameters are disjoint if their result types are disjoint:
+  * `[a1, ..., an] =>> S1 ⋔ [b1, ..., bn] =>> T1` if `S1 ⋔ T1`
+* An `enum` value case is disjoint from any other `enum` value case (identified by either not being in the same `enum` class, or having a different name):
+  * `p.C.x ⋔ q.D.y` if `C != D` or `x != y`
+* Two literal types are disjoint if they are different:
+  * `c ⋔ d` if `c != d`
+* An `enum` value case is always disjoint from a literal type:
+  * `c ⋔ q.D.y`
+  * `p.C.x ⋔ d`
+* An `enum` value case or a constant is disjoint from a class type if it does not extend that class (because it's essentially final):
+  * `p.C.x ⋔ q.D[...Ti]` if `baseType(p.C.x, D)` is not defined
+  * `p.C[...Si] ⋔ q.D.y` if `baseType(q.D.y, C)` is not defined
+  * `c ⋔ q.D[...Ti]` if `baseType(c, D)` is not defined
+  * `p.C[...Si] ⋔ d` if `baseType(d, C)` is not defined
+* Two class types are disjoint if the classes themselves are disjoint, or if there exist a common super type with conflicting type arguments.
+  * `p.C[...Si] ⋔ q.D[...Ti]` if `C ⋔ D`
+  * `p.C[...Si] ⋔ q.D[...Ti]` if there exists a class `E` such that `baseType(p.C[...Si], E) = a.E[...Ai]` and `baseType(q.D[...Ti], E) = b.E[...Bi]` and there exists a pair `(Ai, Bi)` such that
+    * `Ai ⋔ Bi` and it is in invariant position, or
+    * `Ai ⋔ Bi` and it is in covariant position and there exists a field of that type parameter in `E`
+
+It is worth noting that this definition disregards prefixes entirely.
```

Comment:
Disregards _differing_ prefixes?  Just to be clear that not all prefixes are disregarded.